mender-update/standalone/standalone.cpp

```diff
@@ -0,0 +1,516 @@
+		ss << "]";
+	}
+
+	ss << "}";
+
+	string strdata = ss.str();
+	vector<uint8_t> bytedata(strdata.begin(), strdata.end());
+
+	return db.Write(context::MenderContext::standalone_state_key, bytedata);
+}
+
+error::Error RemoveStandaloneData(database::KeyValueDatabase &db) {
+	return db.Remove(context::MenderContext::standalone_state_key);
+}
+
+ResultAndError Install(context::MenderContext &main_context, const string &src) {
+	StandaloneData data;
+	auto exp_in_progress = LoadStandaloneData(main_context.GetMenderStoreDB(), data);
+	if (!exp_in_progress) {
+		return {Result::FailedNothingDone, exp_in_progress.error()};
+	}
+
+	if (exp_in_progress.value()) {
+		return {
+			Result::FailedNothingDone,
+			error::Error(
+				make_error_condition(errc::operation_in_progress),
+				"Update already in progress. Please commit or roll back first")};
+	}
+
+	if (src.find("http://") == 0 || src.find("https://") == 0) {
+		return {
+			Result::FailedNothingDone,
+			error::Error(make_error_condition(errc::not_supported), "HTTP not supported yet")};
+	}
+
+	auto stream = make_shared<ifstream>(src);
+	if (!stream->good()) {
+		int errnum = errno;
+		return {
+			Result::FailedNothingDone,
+			error::Error(
+				generic_category().default_error_condition(errnum), "Could not open " + src)};
+	}
+	io::StreamReader artifact_reader(stream);
+
+	artifact::config::ParserConfig config {
+		paths::DefaultArtScriptsPath,
+	};
+	auto parser = artifact::Parse(artifact_reader, config);
+	if (!parser) {
+		return {Result::FailedNothingDone, parser.error()};
+	}
+
+	auto header = artifact::View(parser.value(), 0);
+	if (!header) {
+		return {Result::FailedNothingDone, header.error()};
+	}
+
+	update_module::UpdateModule update_module(main_context, header.value().header.payload_type);
+
+	auto err =
+		update_module.PrepareFileTree(update_module.GetUpdateModuleWorkDir(), header.value());
+	if (err != error::NoError) {
+		err = err.FollowedBy(update_module.Cleanup());
+		return {Result::FailedNothingDone, err};
+	}
+
+	StandaloneDataFromPayloadHeaderView(header.value(), data);
+	err = SaveStandaloneData(main_context.GetMenderStoreDB(), data);
+	if (err != error::NoError) {
+		err = err.FollowedBy(update_module.Cleanup());
+		return {Result::FailedNothingDone, err};
+	}
+
+	return DoInstallStates(main_context, data, parser.value(), update_module);
+}
+
+ResultAndError Commit(context::MenderContext &main_context) {
+	StandaloneData data;
+	auto exp_in_progress = LoadStandaloneData(main_context.GetMenderStoreDB(), data);
+	if (!exp_in_progress) {
+		return {Result::FailedNothingDone, exp_in_progress.error()};
+	}
+
+	if (!exp_in_progress.value()) {
+		return {
+			Result::NoUpdateInProgress,
+			context::MakeError(context::NoUpdateInProgressError, "Cannot commit")};
+	}
+
+	update_module::UpdateModule update_module(main_context, data.payload_types[0]);
+
+	return DoCommit(main_context, data, update_module);
+}
+
+ResultAndError Rollback(context::MenderContext &main_context) {
```

Comment:
Observation:

Usually you use your whitespace for grouping, but in this function it's kinda random?

Reply:
I think you have to be more specific. It's not random to me.. Just clumping some logical steps together, while separating others which are a bit more unrelated.

Reply:
Maybe I just misunderstood it then :)